Feature: Adding warp-all flag for 0 cost starting Warp

asilverthorn opened the pull request:
Refactor ScanAll class to InitialSpells in the process. This could be used for a future "starting spells" flag set.

Goal with this flag is to support seeds like kaizo, which limit access to warp stones and for which starting Warp Stones do not match the goal of the seed (as they'd likely just be sold off). 

Tested by confirming that the SRAM 0x1a6e - 0x1cf4 look correct:
Just `-scan`:
![scan_all](https://user-images.githubusercontent.com/96998881/169941725-8bda204e-6cc2-4bb9-b58f-2562d58b926b.PNG)

Both `-scan` and `-warp`
![scan_and_warp](https://user-images.githubusercontent.com/96998881/169941726-8416424c-e9de-4f6a-bb3d-91daed388029.PNG)

Just `-warp`:
![warp](https://user-images.githubusercontent.com/96998881/169941728-24d2daf2-3588-431e-8f0f-c4c035dcac57.PNG)

Also tested with these flags and confirmed that all characters started with 0 MP Scan and Warp:
`py wc.py -i ../rom/ff3.sfc -sc1 terra -sc2 locke -sc3 gau -sc4 relm -scan -warp`
![warp_0_cost](https://user-images.githubusercontent.com/96998881/169941732-c106aff1-002d-4bef-93ab-36715a456944.PNG)

asilverthorn committed:
Adding warp-all flag for 0 cost starting Warp. Refactor ScanAll class to InitialSpells

asilverthorn committed:
Adding missed flag to log

kielbasiago commented:
Just got a seed where an esper had warp, would be cool to replace that with another spell if possible (i know this depends on #43)
### Flag List
![image](https://user-images.githubusercontent.com/95580337/170477652-8c1e3f9d-617c-421f-81fd-d78ffbffd474.png)

### Esper Spells
![image](https://user-images.githubusercontent.com/95580337/170477732-350384ad-9674-4793-bf82-2e7d74549647.png)

asilverthorn committed:
Adding Auto Life 3 and Auto Dog Block objective results (#36)

asilverthorn committed:
Stray Flash mod (#33)

asilverthorn committed:
Adding Osteoclave's bugfix (#29)

asilverthorn committed:
Adding expensive rods and super balls flags (#27)

asilverthorn committed:
Bugfix: Fix weapon special abilities with Capture and Multi-steals only giving 1 item (#25)

* Porting Assassin's Capture bugfix

* Adding Bropedio's Multi-Steal fix

* Adding multi-steal dialog fix

* updating flag doc

* Small reduction in C1 usage

* Fixing bug with -fc that causes escape from battle (and possibly other actions) to crash

asilverthorn committed:
Feature: Flashes: Moving flashes to Graphics; Slowing scrolling background during Air Force fight and Baren falls; add minimap high contrast option (#23)

* Removing scrolling background during Air Force fight

* switching to slowing the bg scroll. Also adding baren falls

* Removing baren falls background vibration

* Fixing typo in baren falls background palette change for frm

* Changing events flashes from no-ops to flash-none to avoid any CPU cycle changes

* Moving Remove Flashes flags from misc to graphics, so that they don't affect the seed

* Adding -wmhc option to improve visibility of minimap

* Making location indicator for wmhc stand out even more

* Adding more event flashes to frw

* Changing wmhc based on color feedback

* Replacing boss death flash removal with flash of monster sprite

* switching to Osteoclave's pixel remaster inspired minimap

asilverthorn committed:
Feature: Adding flags to improve Sketch/Control (#19)

* Adding flag to make Sketch/Control 100% effective against valid targets

* fixing minor typo

* Refactoring data/sketch_control.py into separate files. Adding better abilities for Sketch & Control. Making Sketch use caster's stats.

* Minor change to option logging

* Modification to sketch custom command; split into multiple flags; removing control custom commands

* Expanding sia to also include control. Separating Controls from Coliseum monster table.

* Minor arg desc change

* Control uses controller's stats, fix Control bug with Dance moves

* Removing ultima from sketch magimaster

asilverthorn committed:
Feature: Kielbasiago's Adding flag for giving starting Espers (#18)

* add --starting-espers-random flag
- this will give you between the selected range of espers from the very start

* fix err

* Add better esper log

* chore: update flag from sespr => stesp

* chore: remove commented code

* Remove "Random" from "Starting Espers" menu

* Update starting_espers MAX to account for logic issues

* ws, help updated

Co-authored-by: Kiel <95580337+kielbasiago@users.noreply.github.com>

asilverthorn committed:
Feature: Kielbasiago's Update ability to shuffle/mix both dragons and statues (#17)

* add -bmbs flag: adds statues to shuffle pool

* add context for statue ids being in EnemyPacks

* set default true ffor backwards compatibility

* remove default as no way to turn it off

* wip refactor shuffle to "boss location" flags

* bug fixes

* add boilerplate for adding future checks

* chore: no longer appnd dragon locations to flags automatically

* chore: add EOL

* chore: remove unused ids from EnemyPacks

* remove dbugpy code

* Clean up       enemy_packs.randomize_event_bosses

* fix when original bosses with mixed dragon/statues

* now exclude statues when not mix from normal encs

* add KT bosses to condition checks (#2)

Preview: https://youtu.be/wgBJHPq3p6o (Just me clearing KT while playing at 200-400%)
 
Added the following KT boss locations as objective conditions:
- Kefka's Tower Ambush
  - Inferno
  - Bit 59
- Kefka's Tower Guardian
  - Guardian
  - Bit 60
- KT Left Triad Statue
  - Doom
  - Bit 61
- KT Mid Triad Statue
  - Poltergeist
  - Bit 62
- KT Right Triad Statue
  - Goddess
  - Bit 63


## Testing
Used the following flags to test the five encounters: `-oa 40.1.1.11.59 -ob 40.1.1.11.60 -oc 40.1.1.11.61 -od 40.1.1.11.62 -oe 40.1.1.11.63`

* fix initialization of exclude_bosses

Co-authored-by: Nolan <10077353+nolanlocke@users.noreply.github.com>
Co-authored-by: Kiel <95580337+kielbasiago@users.noreply.github.com>

asilverthorn committed:
Feature: Magic MP Randomization (#15)

* Adding Ultima 255 MP flag, Adding Franklin's Magic MP Randomization, Fixing display bugs with 3 digit MP, letting Lore MP randomization go to 255 MP

* Allowing random esper MP cost to go up to 255

* Switching max to 254 to avoid boundary conditions

* making 254 max to avoid issues with MP checks

* Moving from c3 to f0

asilverthorn committed:
Adding --starting-level flag (#14)

asilverthorn committed:
QoL: Adding Rage Move descriptions to rage menu (#13)

* Initial commit of Rage ability names in menu

* Adding logic to display Special rage names

* Adding logic for Special (0xef) attacks

* Rewrite of Rage Description to use custom string table that includes more move details

* More details in rage desc; handling special cases

* Simplifying rage strings

asilverthorn committed:
Feature: Add Remove learnable spells flag + Submenus (#12)

* All original remove learnable spells changes

* Making Life 3 learnable in permadeath seeds

* Fixing bug where Natural Magic users can learn Fire even if excluded

* correcting natural magic learning with excluded commands

* Fixing seed gen error when there's not enough spells available for randomization

* Fixing possibility for long-branches with many submenus

asilverthorn committed:
Removing excluded non-S tier items from tiered or scaled chests (#11)

asilverthorn committed:
Ensuring that Gau can use Magic in FT (#10)

asilverthorn committed:
Feature: Adding random-encounters-chupon flag (#8)

* Adding random-encounters-chupon flag

* Making Chupon 64 sneeze always target entire party

asilverthorn committed:
Feature: Kielbasiago's add --start-junk flag (#7)

* add --start-junk flag

* Add relics to starting junk

Co-authored-by: Kiel <95580337+kielbasiago@users.noreply.github.com>

asilverthorn committed:
QoL: Setting default config options to most commonly used values (#6)

* adding comments

* setting default back to Reset and adding Config 2 relocation for compatibility with DoctorDT tool

asilverthorn committed:
Making Party warp to Arvis house following K@N (#4)

asilverthorn committed:
Feature: Adding -llr flag to randomize L.x lore levels (#3)

* Adding -llr flag to randomize L.x lore levels

* Updating lore battle dialogs

asilverthorn committed:
QoL: Adding clarification of quantity of objective conditions required (#1)

* Adding clarification of quantity of objective conditions required

* responding to feedback on objective menu mods

ff6wc committed:
Update version.py

v1.1.0 -> v.1.2.0

ff6wc committed:
Update flags.py

Updating to match dev branch

ff6wc committed:
Update spells.py

Updated to match dev branch

asilverthorn committed:
Feature: Making Top 4 Magitek commands an objective result (#21)

* Making Terra's Magitek commands an objective result

* Minor changes to address PR comments

Co-authored-by: Will Jones <wrjones104@users.noreply.github.com>

wrjones104 committed:
Merge branch 'v1.2.0' into warp_all

asilverthorn commented:
Merged into ff6wc fork
